File: cplus/leetcode/202502/algo0377.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 - 爬楼梯
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        // 不能用背包，背包组合数不计算顺序
        // 普通爬楼梯 1、2 步
        // 本题相当于可用 nums 中方式爬上 target 高的楼梯
        vector<unsigned> f(target+1); // int 会溢出
        f[0] = 1;
        for (int i=1; i<=target; i++) 
            for (auto& v:nums) 
                if (i >= v) 
                    f[i] += f[i-v];
        return f[target];
    }
};

File: python/algo/202502/algo0377.py
from typing import List

# 动态规划 - 爬楼梯
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0] * (target + 1)
        dp[0] = 1
        for i in range(1, target + 1):
            for v in nums:
                if i >= v: dp[i] += dp[i-v]
        return dp[target]